docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,88 @@
+* There is no application specific way to tell if the call successfully executed. For example, the canister does not provide a way to find out if the call happened.
+
+In this situation, it is indistinguishable if the call was successful, but all information has already been removed from the system state tree, or the call never made it into the state tree in the first place and hasn't been executed. 
+
+### Recommendations and best practices for secure update call handling
+
+The best way to handle the aforementioned errors is to implement safe retries using idempotent calls. Idempotency and several patterns to achieve it are discussed in the [best practices on safe retries and idempotency](/docs/current/developer-docs/smart-contracts/best-practices/idempotency).
+
+It may not always be possible to make call endpoints idempotent. For example, an endpoint that is not under the developer’s control may simply not have this property. In this case, there is still a secure way to retry the calls by relying on built-in deduplication based on request ids, as we describe below. However, the guarantees in this case are weaker compared to using idempotent calls.
+
+ICP provides a built-in deduplication feature. Namely, if two calls with identical [request id](/docs/current/references/ic-interface-spec#request-ids) are made, the message is processed at most once. For two calls to have the same `request_id`, they must have the same values for all fields `request_type`, `sender`, `nonce`, `canister_id`, `method_name`, `arg` and **ingress_expiry**. This gives idempotency, but only until the [`ingress_expiry`](/docs/current/references/ic-interface-spec#authentication) is reached, at which point the message expires and ICP [may remove the `request_id` from the system state tree](/docs/current/references/ic-interface-spec#state-tree-request-status). This isn't an issue since the same request will not be accepted later, as it expired. Thus, such requests can be safely retried analogously to how it is done for [idempotent calls](/docs/current/developer-docs/smart-contracts/best-practices/idempotency#idempotent-canister-apis), see the illustration on that page.
+
+As ICP rejects `ingress_expiry` that is too far in the future, the time window for retrying is limited. To maximize the window, one can maximize `ingress_expiry` in this case and time out the retries after `ingress_expiry` as well.
+
+There are two options for retries:
+* Submit the `call` once, and poll `read_state` while at least one of the following conditions is met: 
+  * `ingress_expiry` is not reached.
+  * The call status is `processing`.
+
+  Evaluating the first condition may be hard if the IC time and local time differ. Currently, agent-js and agent-rs by default poll until `ingress_expiry` is reached. 
```

Comment:
```suggestion
  Evaluating the first condition may be hard if the subnet time and local time of the user differ. Currently, agent-js and agent-rs by default poll until `ingress_expiry` is reached. 
```
since every subnet can have a different time, it's better to use subnet time here.